constexpr typeid improvements

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26482 [wip] Error checking for KernelFunction::callUnboxed()
* #30712 Better error messages thanks to compile time type names
* #31307 [no-land] try fix 2
* #31306 [no-land] try fix 1
* #26628 Remove CAFFE_KNOWN_TYPE
* #31313 Fix c10::util::get_fully_qualified_type_name for MSVC
* #31351 Make c10::util::get_fully_qualified_type_name() backwards compatible with clang 4
* **#31312 constexpr typeid improvements**

Differential Revision: [D19087198](https://our.internmc.facebook.com/intern/diff/D19087198/)

smessmer committed:
[wip] constexpr typeid improvements



Differential Revision: [D19087198](https://our.internmc.facebook.com/intern/diff/D19087198/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] constexpr typeid improvements"


Differential Revision: [D19087198](https://our.internmc.facebook.com/intern/diff/D19087198/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit c19a9f7:
* **2/2** broken upstream at merge base 930d075 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=930d0751e6a3006c6251c095ead4dcab71b83d52&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=c19a9f7cdacc2a04235a72df95d1bd9d39fdd19b).

### 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py2_gcc5_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75112735)
* [caffe2_onnx_py3_6_clang7_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75113385)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer committed:
Update on "constexpr typeid improvements"


Differential Revision: [D19087198](https://our.internmc.facebook.com/intern/diff/D19087198/)

[ghstack-poisoned]

dzhulgakov commented on c10/util/typeid.h:
```diff
@@ -454,12 +454,12 @@ inline TypeMeta::TypeMeta() noexcept
 }
 
-inline constexpr bool operator==(
+inline bool operator==(
```
why did we remove it here?

smessmer replied:
because this wasn't truly constexpr. `TypeMeta` objects can't even be constructed in a constexpr way currently. I might find a way to do it in future PRs.

smessmer committed:
Update on "constexpr typeid improvements"


Differential Revision: [D19087198](https://our.internmc.facebook.com/intern/diff/D19087198/)

[ghstack-poisoned]
